### ISSUE TITLE
Next steps

### ISSUE THREAD

**Issue by ghost** [ikvm-revived/ikvm#13]
@timotei, @ams-tschoening, @Draspax - do you three want to start in on bringing in the changes for .Net Core (.Net Standard 2.0)?

I'll ask @igorkwwrd if he can add:
1. StrongNaming - a .snk file we have that is not checked in. And hopefully a way to build with a checked in key for others.
2. The latest SharpZipLib (needed to be .Net Standard 2.0 compliant)
3. Put in NuGet - as distinct DLLs. @jfrijters - can we keep the IKVM name? If we change it people will not be able to find it because that's what they search on.

??? - thanks - dave

**Comment by igorkwwrd** [comment]
Is it fine to have an .snk file checked in?

ZipLib upgrade would be a part of the migration to the core.  It depends on what we chose for the build tools.  Currently it's NAnt which looks unmaintained for a while.
I think msbuild would be a natural choice which allows building on Windows & Linux the same way.

If there are no objections I would stay with the IKVM name.

**Comment by DavidThi808** [comment]
@igorkwwrd I agree on switching to MSBuild, upgrading ZipLib, and keeping the IKVM name.

I think we should not check in the .snk file we use because it would allow anyone to create a hacked version of the DLLs and substitute them in???

**Comment by jfrijters** [comment]
Keeping the name is fine.

**Comment by DavidThi808** [comment]
@jfrijters Thank You

**Comment by NightOwl888** [comment]
It occurred to me that I have never seen a configuration using a combination of the old `.csproj` format and a .NET Standard/.NET Core target, so I decided to run an experiment by creating an old .NET Framework library and then editing the project file to select the `netstandard2.0` moniker. As I suspected, using the old format with .NET Standard or .NET Core is not possible.

The old library contains a section like this:
```
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{077BE7D3-3EF5-4A6B-9198-274EC04FBD08}</ProjectGuid>
    <OutputType>Library</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>ClassLibrary1</RootNamespace>
    <AssemblyName>ClassLibrary1</AssemblyName>
    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
  </PropertyGroup>
```

Since `<TargetFrameorkVersion>` only supports .NET Framework versions, I tried just swapping it to the new `<TargetFramework>` element:

```
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{077BE7D3-3EF5-4A6B-9198-274EC04FBD08}</ProjectGuid>
    <OutputType>Library</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>ClassLibrary1</RootNamespace>
    <AssemblyName>ClassLibrary1</AssemblyName>
    <TargetFramework>netstandard2.0</TargetFramework>
    <FileAlignment>512</FileAlignment>
  </PropertyGroup>
```

But that doesn't work because `<TargetFramework>` is not defined in the old XML schema. The main issue is that .NET Standard now uses Roslyn (which uses MSBuild properties) to compile the library, and to do that, you have to change the `<Project>` element from:

```
<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```

to

```
<Project Sdk="Microsoft.NET.Sdk">
```

But there are other issues to resolve when converting from the old format to the new format. There are a few guides on how to do so, such as [Old csproj to new csproj: Visual Studio 2017 upgrade guide](https://natemcmaster.com/blog/2017/03/09/vs2015-to-vs2017-upgrade/), [Upgrading Old csproj to New csproj in Visual Studio 2017](https://amanagrawal.blog/2018/07/02/upgrading-old-csproj-to-new-csproj-in-visual-studio-2017/), and [Upgrading an existing .NET project files to the lean new CSPROJ format from .NET Core](https://www.hanselman.com/blog/UpgradingAnExistingNETProjectFilesToTheLeanNewCSPROJFormatFromNETCore.aspx). There is also a tool to automate the conversion [here](https://github.com/hvanbakel/CsprojToVs2017).

Personally, I am a fan of the "Start Over" approach, as it is the best way to ensure you pare the XML down to the minimal configuration that you require.

The bottom line is that switching to the new `.csproj` format as [I had previously suggested](https://github.com/ikvm-revived/ikvm/pull/12#discussion_r382943674) is not optional to support .NET Standard 2.0, **it is a requirement**.

And since that format is setup to automatically generate both `AssemblyInfo.cs` and `.nuspec` files, there is no reason to deal with parameterization in those formats, you only need to create the parameters in the MSBuild format and pass them in from the build pipeline. It is still possible to do it like this, but you have to manually suppress the automated generation to avoid conflicts.

I'd be happy to contribute getting the project setup for .NET Standard 2.0, but I am not familiar with the Ant format it is currently using. If someone were to do the work "Convert Ant to MSBuild" (on .NET Framework), I would be happy to complete the requirement "Convert MSBuild to Roslyn", which would get us building xplat and compiling for .NET Standard 2.0/.NET Core.

**Comment by igorkwwrd** [comment]
I think I can do Ant -> Msbuild (framework). There are some .csproj files already. I'll follow to make it all in sync with Ant's .build files.

**Comment by NightOwl888** [comment]
Okay, great. 

Please just do the minimum to get it to build (ideally both in VS and on the command line). Don't worry about version number incrementing, strong naming, NuGet packing, etc. at this point. That can all be parameterized after upgrading to the new format, and then any CI tool can be used to pass in the parameters.

**Comment by DavidThi808** [comment]
Hi all;

There's four projects in IKVM that have issues. Forms is a gigantic number, so I think we wait on that (and we don't need it). JDBC is easy but again I don't know if anyone uses it so wait there.

That means we have just 23 items, 17 of them reflection. So anybody have any parts they want to sign up for? **And has anyone already done some of these?**

I'll take all the non reflection ones if someone else can do the reflection ones.

??? - thanks - dave

[IKVM_Port_to_dotnet_standard.xlsx](https://github.com/ikvm-revived/ikvm/files/4273033/IKVM_Port_to_dotnet_standard.xlsx)

**Comment by NightOwl888** [comment]
#### `M:System.IO.FileStream.#ctor(System.String,System.IO.FileMode,System.Security.AccessControl.FileSystemRights,System.IO.FileShare,System.Int32,System.IO.FileOptions)`

The overload only applies to the Windows OS. There is another overload available in .NET Standard 2.0, `M:System.IO.FileStream.#ctor(System.String,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,System.Int32,System.IO.FileOptions)` that is cross-platform. Can we use this instead?

An [issue that was opened about this](https://github.com/dotnet/runtime/issues/30435#issuecomment-590609103) says that the overload has been added in .NET Core 3.1, which is only helpful if we plan to multi-target.

#### `Reflection.Emit`, `System.AppDomain.DefineDynamicAssembly`, `System.ArgIterator` 

The functionality can probably be replaced with Roslyn. See [Convert Reflection.Emit to Roslyn](https://stackoverflow.com/q/38067666). This is going to take some elbow grease to do, it is not a 1-to-1 conversion.

#### `M:System.Diagnostics.StackTrace.#ctor(System.Threading.Thread,System.Boolean)`

This looks like it might be an issue with the analyzer tool, as I don't see this overload being used in the codebase. In `runtime/vm.cs`, it is using the overload `M:System.Diagnostics.StackTrace.#ctor(System.Exception,System.Boolean)` which may be tripping up the analyzer tool. This overload exists in .NET Standard 2.0.


### Suggestion

After the project files are upgraded to .NET Standard, code that is not compiling can be fenced off using conditional compilation. For example:

```c#
#if !NETSTANDARD

// Code that is not compatible with .NET Standard

#else

// Stub to make the code compile

#endif
```

Then the work on the build can be finished without worrying about if we have all of the pieces in place. Once the build is done, the parts that are missing (mostly `Reflection.Emit`) can be converted in sections to be compatible with .NET Standard. 

The `Reflection.Emit` code could be left in place to be able to use a .NET Framework target to compare the Roslyn conversion to ensure the result is the same.

**Comment by ams-tschoening** [comment]
>  So anybody have any parts they want to sign up for?

I most likely won't have the time to commit to anything concrete, priorities have shifted recently.

> And has anyone already done some of these?

[net_core_compat](https://github.com/ikvm-revived/ikvm/commits/net_core_compat) should work around at least some of the problems related to FileStream-CTORs. This should be reviewed before merging if they are worth merging at all. Sometimes appends can be implemented using different methods, sometimes appends are the only things not working, so I throw exceptions in those cases and allow others. I've introduced TODOs to mark things as not compatible:

> // TODO NET_CORE_INCOMPAT

What I additionally recognized was that getting things to work using DEBUG doesn't tell much: Things need to additionally work using RELEASE and the native compiler toolchain and there are additional problems with those I documented as issues and PRs #4 and #5 already.

**Comment by DavidThi808** [comment]
> which is only helpful if we plan to multi-target.

My preference is we try to do this without multi-targeting. But if we do need to multi-target, we limit it to NETFRAMEWORK, NETCORE_3_1, and NETSTANDARD_2_0. It's not worth the hassle to do other builds for the 1 or 2 items added.

**Comment by DavidThi808** [comment]
> The functionality can probably be replaced with Roslyn. See Convert Reflection.Emit to Roslyn. This is going to take some elbow grease to do, it is not a 1-to-1 conversion.

@NightOwl888 can you do this part?

**Comment by DavidThi808** [comment]
My one concern is - are these methods needed? If so, we should do an #if NETFRAMEWORK around the old way.

**Comment by DavidThi808** [comment]
> Things need to additionally work using RELEASE and the native compiler toolchain and there are additional problems with those I documented as issues and PRs #4 and #5 already.

I think those changes (#4 & #5) should use:
#if NETFRAMEWORK
... old way ...
#else
throw UnsupportedOperationException()
#endif

**Comment by NightOwl888** [comment]
>> The functionality can probably be replaced with Roslyn. See Convert Reflection.Emit to Roslyn. This is going to take some elbow grease to do, it is not a 1-to-1 conversion.

> @NightOwl888 can you do this part?

Possibly. I am not sure if I will have the time to.

I have used Roslyn only to make a [code analysis rule and fix](https://github.com/apache/lucenenet/tree/master/src/dotnet/Lucene.Net.CodeAnalysis), but not for generating dynamic assemblies.

**Comment by DavidThi808** [comment]
Maybe the first approach should be use other .Net calls where we can. And on the ones we can't, throw an exception.
Because we don't know if this code normally gets called.

**Comment by ams-tschoening** [comment]
What's the rules for multi-targeting? Do we really keep everything already available targeting .NET Framework and add support for .NET Standard only using `ifdefs` or is it acceptable to remove old code not compatible with .NET Standard? I simply followed the latter approach in `net_core_compat`, because I think the code base is already pretty cluttered with `ifdefs`.

But .NET Framework is still needed by WWRD and others?

https://docs.microsoft.com/de-de/dotnet/standard/net-standard

**Comment by NightOwl888** [comment]
> What's the rules for multi-targeting? Do we really keep everything already available targeting .NET Framework and add support for .NET Standard only using ifdefs or is it acceptable to remove old code not compatible with .NET Standard? I simply followed the latter approach in net_core_compat, because I think the code base is already pretty cluttered with ifdefs.

> But .NET Framework is still needed by WWRD and others?

A .NET Standard 2.0 library will be able to be consumed by any .NET Framework project 4.6.1 or higher. Technically, a .NET Framework target is only required if we need to support .NET Framework < 4.6.1. Keep in mind, Microsoft no longer supports anything before .NET Framework 4.5.2, and that support [will end in 2023](https://stackoverflow.com/a/59653801).

That being said, I think it should be done in the interim period while we get things compiling, because it is useful to be able to compare the output of each to make sure our conversion is producing the same result. Once we have completed that, we can (optionally) remove the .NET Framework target from the official build.

**Comment by DavidThi808** [comment]
Windward absolutely needs a .NET Framework version as do many others out there. We're 4.6.1 but we use things like OData versions 1 - 3 which has libraries for .Net Framework but not .Net Standard or .Net Core.
This is the beauty of NuGet. We can push up 3 sets of DLLs, one each for .Net Standard 2.0, .Net Core 3.1, & .Net Framework. When someone uses NuGet for their app, it'll pull the one that matches.

**Comment by NightOwl888** [comment]
> Windward absolutely needs a .NET Framework version as do many others out there. We're 4.6.1 but we use things like OData versions 1 - 3 which has libraries for .Net Framework but not .Net Standard or .Net Core.

Actually, the fact you need OData makes no difference. An assembly that targets .NET Framework 4.6.1 can simultaneously consume a .NET Standard 2.0 targeted library and a .NET Framework targeted library. So, unless I am missing something, this isn't going to be a limitation in your case at all. .NET Standard is simply a façade layer that under the covers calls .NET Framework APIs when the consuming application targets .NET Framework.

I agree that the plan should be to use a single target, and only multi-target if we find a limitation we can't work around.

**Comment by DavidThi808** [comment]
> An assembly that targets .NET Framework 4.6.1 can simultaneously consume a .NET Standard 2.0 targeted library and a .NET Framework targeted library.
You're correct. I used that as an example. A better example might be if it's a better solution to use the existing non-supported FileStream constructors. We then provide a sub-standard solution under Standard and the complete solution under Framework.
Best is if we can go 100% standard. But let's keep the ability to build distinct systems if needed.

**Comment by NightOwl888** [comment]
I started going through what it would take to make the assemblies from the IKVM NuGet package into .NET Standard 2.0 assemblies. However, being that most of them are generated using scripts, it means that we can't just target `netstandard2.0`, build, and deploy - the scripts will also need to be altered.

The list of end user assemblies is as follows:

- IKVM.AWT.WinForms.dll
- IKVM.OpenJDK.Beans.dll
- IKVM.OpenJDK.Charsets.dll
- IKVM.OpenJDK.Cldrdata.dll
- IKVM.OpenJDK.Corba.dll
- IKVM.OpenJDK.Core.dll
- IKVM.OpenJDK.Jdbc.dll
- IKVM.OpenJDK.Localedata.dll
- IKVM.OpenJDK.Management.dll
- IKVM.OpenJDK.Media.dll
- IKVM.OpenJDK.Misc.dll
- IKVM.OpenJDK.Naming.dll
- IKVM.OpenJDK.Nashorn.dll
- IKVM.OpenJDK.Remoting.dll
- IKVM.OpenJDK.Security.dll
- IKVM.OpenJDK.SwingAWT.dll
- IKVM.OpenJDK.Text.dll
- IKVM.OpenJDK.Tools.dll
- IKVM.OpenJDK.Util.dll
- IKVM.OpenJDK.XML.API.dll
- IKVM.OpenJDK.XML.Bind.dll
- IKVM.OpenJDK.XML.Crypto.dll
- IKVM.OpenJDK.Parse.dll
- IKVM.OpenJDK.Transform.dll
- IKVM.OpenJDK.WebServices.dll
- IKVM.OpenJDK.XPath.dll
- IKVM.Runtime.dll
- IKVM.Runtime.JNI.dll
- x64/ikvm-native-win32-x86.dll
- x64/JVM.dll
- x86/ikvm-native-win32-x86.dll
- x86/JVM.dll

There are a few hurdles to get over before these can be made. Here are the first 2:

1. It turns out that while IKVM.Reflection can be directly converted to .NET Standard 2.0 and executes fine, this library is not part of the finished product. It is used in generation of JVM.dll and IKVM.Runtime.dll. However, after doing some preliminary research, [Reflection.Emit cannot generate .NET Standard libraries](https://github.com/dotnet/standard/issues/884#issuecomment-430815440). So, it looks like IKVM.Reflection will need to be rewritten using Roslyn syntax nodes after all in order to generate the deliverables JVM.dll and IKVM.Runtime.dll.
2. The IKVM.OpenJDK dlls are also generated. I haven't worked out exactly which component is responsible for doing this (it is apparently related to the `openjdk\response.txt` file), but it will also need to be changed to generate .NET Standard 2.0 libraries (or failing that, .NET Core 2/3 libraries). If `CSC.exe` is called somewhere to generate those assemblies, it should be just a matter of replacing those `CSC.exe` calls with `dotnet build` or `dotnet msbuild` and fixing any incompatible API calls in the sources.

It would definitely be advantageous to put the DLLs into separate NuGet packages, since some features (such as WinForms) won't be available unless .NET Framework is targeted. Being able to choose different targets for each assembly makes this easier to manage.

Also, it occurred to me that since there are so many command line utilities, it might be best to create a [dotnet tool](https://github.com/natemcmaster/dotnet-tools) to encapsulate all of these different commands into a single utility, which can be [delivered via NuGet](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-install#examples).

Anyway, #17 is ready for review.

**Comment by DavidThi808** [comment]
I was using Api Portability to determine what needs to change. But I've learned it gives false positives ([details here](https://stackoverflow.com/questions/60600311/how-can-i-determine-what-libraries-are-not-good-on-net-core)). So just because ApiPort says a package is not supported, that may be incorrect.

**Comment by NightOwl888** [comment]
> I was using Api Portability to determine what needs to change. But I've learned it gives false positives (details here). So just because ApiPort says a package is not supported, that may be incorrect.

The portability analyzer doesn't apply in our case because it is designed to analyze static code, and the assemblies that are published to IKVM consumers are (for the most part) dynamically generated.

**Comment by igorkwwrd** [comment]
I encountered a problem with ikvmstub. It generates a jar file from a given .net assembly. E.g.
ikvmstub mscorlib

It's all good with the .net framework. Its mscorlib.dll is ~5M in size. It contains namespaces, types, etc.

However, there is a problem with the .net core. Its mscorlib.dll is ~50K and contains a manifest only.
This prevents me from generating a corresponding jar file. Which in turn prevents compiling the openjdk project.

Any ideas?
Maybe there is a way of reading that manifest and load all assemblies listed in it?

**Comment by NightOwl888** [comment]
I am not sure whether this will be helpful, but according to [this comment](https://github.com/dotnet/standard/issues/884#issuecomment-430895301), when generating a .NET Standard library (which is what we should be aiming to make ikvm do), you must manually link to the correct version of netstandard.dll (see the other comments for locations where different versions of it can be found).

If the problem happens to be at a higher level (the IL generation), I noticed that the [OpCode.cs file in `IKVM.Reflection` has a `Main()` method for testing the CIL that is generated](https://github.com/ikvm-revived/ikvm/blob/master/reflect/Emit/OpCode.cs#L151) using `System.Reflection.Emit` vs `IKVM.Reflection`. `IKVM.Reflection` is a dependency of `ikvmstub`.

Also, it is not out of the question that something is going haywire because of the new mismatch between the target platform for .NET Core executables consuming libraries that are targeting .NET Standard (since as I recall, we are targeting net461 in both cases where it is succeeding). As an experiment, it might be helpful to temporarily target .NET Core in both executables and libraries to see if the entire problem is being caused by this mismatch.

**Comment by DavidThi808** [comment]
I think we should do this with several criteria for now (we can expand in the future):

1. The compiler runs on .Net Framework. The generated DLLs are .Net Core or .Net Standard. But the compiler requires .Net Framework. If System.Reflection.Emit is only needed by the compiler, then it's no longer an issue.

1. If .Net Core 3.1 is a lot easier than .Net Standard 2.0, then lets accomplish .Net Core first. That is a giant win. We can then do .Net Standard next.

1. Let's get the parts needed for a server app first. So we can skip Swing, JDBC, etc. We do those next, but a lot of apps (Windward, Saxon, etc.) just need the core part. That gets those running.

I think we need success along the way to make this successful.

**Comment by igorkwwrd** [comment]
A .net core DLL contains a forwarded types concept which is unknown to the
.net framework.

Many assemblies in the .net core contain just forwarded types that reference
other assemblies.

IKVM.Reflection needs to be changed to read in the missing forwarded types
table to make those types available at runtime.

A modular nature of the .net core is another issue.  When you access a
forwarded type its assembly needs to be available on a host.  Otherwise,
you get a file not found exception.

A generic solution could be to dynamically download missing assemblies from
nuget.  As a workaround we can reference all needed assemblies at the project
level.  This ensures that all assemblies are availabe at the time we build
the project and then execute the tools we need, ikvmstub, ikvmc, etc.

There are some things we need in System.Reflection.Emit that exist in the
framework but are missing from the core.  Some of them are used to emit
a debug information, so they may be skipped for now.  Some others might be
essentials, like mark position, tokens, etc. (from the AssemblyBuilder).

Making the ikvm compiler .net framework based and try to produce .net core
assemblies is not an option.  It won't be able to produce a correct .net
core assembly, for example due to the mentioned problems.

There might be other incompatibilities I'm not aware of yet.

So, we have not simply "build it right way" problem, but rather quite lot of
work involving code change.

**Comment by NightOwl888** [comment]
Thanks Igor, I was trying to work out how to explain that, and some of it (such as new features that are missing from .NET Framework) I was only suspecting, so thanks for the confirmation.

> A modular nature of the .net core is another issue. When you access a
forwarded type its assembly needs to be available on a host. Otherwise,
you get a file not found exception.

> A generic solution could be to dynamically download missing assemblies from
nuget. As a workaround we can reference all needed assemblies at the project
level. This ensures that all assemblies are availabe at the time we build
the project and then execute the tools we need, ikvmstub, ikvmc, etc.

The new build chain generates a dependency graph in the `obj` folder. The main entry point is the `project.assets.json` file. The `dotnet restore` command generates this file and then uses the graph to download the dependencies. This is how it is typically set up on CI environments to install/locate the dependencies, which may be NuGet packages or (in the case of .NET Framework), local paths.

There are at least 2 avenues we could pursue:

1. Utilize `dotnet.exe` directly to compile the assemblies that are currently generated using utilities such as `ikvmstub.exe` to leverage the existing functionality that resolves dependencies.
2. Reverse engineer `dotnet.exe` (since it is open source) and copy the bits that build the dependency manifest and do the restore into IKVM.

In either case, I suggest having a look at the [available commands in dotnet.exe](https://docs.microsoft.com/en-us/dotnet/core/tools/). There are several new commands that weren't available with prior build tools that we might be able to leverage.

**Comment by DavidThi808** [comment]
This may be a dumb question but... what if we converted the created .Net Framework DLLs to .Net Core? In other words leave IKVM alone and have a post processor?

It would require we don't use any classes that don't exist in .Net Core, but otherwise can this be done?

**Comment by igorkwwrd** [comment]
Some good news after debugging

ikvm.reflection sees forwarded types and there is nothing to change there
ikvmstub has the '-forwarders' flag.  Using that I could produce a fully populated mscorlib.jar
(how didn't I see the flag before)

**Comment by igorkwwrd** [comment]
IAsyncDisposable is one more thing that exists in the core but not in the framework.

**Comment by igorkwwrd** [comment]
Parts of System.Data are not available entirely or partially in the core, e.g. Microsoft.SqlServer, System.Data.Odbc, System.Data.OleDb, System.Data.SqlClient.

This led to removing the sun/jdbc/odbc namespace from the java sources as they can't be compiled without the complete System.Data.

So we won't be producing IKVM.OpenJDK.Jdbc.dll for some period in the future.

**Comment by DavidThi808** [comment]
I think using JDBC through IKVM is a poor decision regardless. Use ADO.NET for that part. So I'm good with not getting IKVM.OpenJDK.Jdbc.dll ever.

**Comment by NightOwl888** [comment]
> Parts of System.Data are not available entirely or partially in the core, e.g. Microsoft.SqlServer, System.Data.Odbc, System.Data.OleDb, System.Data.SqlClient.

Actually, they are through .NET Platform Extensions. You just need to install the NuGet package to use them, they are no longer part of the core assembly. For example, [System.Data.SqlClient](https://www.nuget.org/packages/System.Data.SqlClient/).

**Comment by klokedm** [comment]
Hi all,

based on the #17 I've created a .NET Core / Standard only build to make it easier for me to diagnose dependency issues.

The results are here: https://github.com/klokedm/ikvm (master branch) and are as follows:
  - the build process builds everything using .NET Standard 2.0 &and 2.1 for IKVM.Runtime, as it depends on some unavailable methods in 2.0) + Platform Extensions
  - the ikvmc build using .NET Core 3.1 + Platform Extensions
  - some minor modifications were required to the IKVM openjdk java classes to be compatible with the new File System API
  - I removed some code required for emitting debug assemblies as it used incompatible API from the compiler
  - I added stub geeneration for required platform extensions for the javac to be able to build the openjdk
  - I replaced the mscorlib references with Private.CoreLib in the compile process, this probably needs to be changed further to correct forwarders.
  - I changed the assembly comparisson code to work with the Platform Extensions; this needs to be further reviewed

NOTE: due to different projects targetting different frameworks, some paths are currently hardcoded in the build file.

If you want to try out the above branch, just unzip the "openjdk-8u45-b14" into the openjdk folder and run "dotnet build --configuration Release"

Everything builds up to the point of the ikvmc trying to package and build the javac compiled .class files to IKVM.OpenJDK packages. There, the compiler fails to resolve the System.Attribute class when resolving it in the IKVM.Runtime. I did not have time yet to study the compiler mechanics as to why this happens after the switch / changes.

Feel free to use any of the changes / code you find useful and let me know if I can help with anything specific.

**Comment by klokedm** [comment]
A short update:
 - The ikvmc now uses netstandard as the runtime
 - After a few changes to the build script and the compiler, the OpenJDK now compiles with the .NET core version of the compiler taregtting the netstandard.dll (instead of mscorlib)
 - also, an update to the openjdk map file was needed to remove the reference to mscorlib and replace it with Private.CoreLib

The current version of the compiler has a quick fix (as in probably a nasty hack) in type verification to allow for the same type imported via different routes to be used - it compares the types by name instead of by wrapper reference. This needs to be further researched and resolved.

The main problem with the build was the assembly original resolver has been designed in a way to expect all assemblies, including the ones the functions are forwarded to need to be specified to be loaded in the command line; I've updated the compiler to output a missing assembly, if encountered to make diagnostics easier and will probably create an updatd loader to automatically handle those dependencies.

Long story short: the compiler now *builds* on the .NET core stack and is able to produce the OpenJDK libraries that are linked to .NET Standard dll, however they need to be tested if they are usable at all.

**Comment by NightOwl888** [comment]
@klokedm 

Nice work. I suggest creating a PR so these changes get integrated. Perhaps until the review process is worked out, just target the #17 branch so it is integrated more quickly.

**Comment by NightOwl888** [comment]
Since it seems to be not possible to get a PR into the master branch at the moment, I suggest the team look into the flaws in the current review procedure so progress can be made on IKVM without having to wait several weeks for something to happen. There is nothing wrong with requiring 4 people in review the process as long as each of them understands their role, but it appears some do not.

1. Notifications should be sent to the reviewers informing them what their responsibilities are in the review process.
2. Perhaps a document should be created to ensure this procedure is not forgotten.
3. The procedure should be clear as to what the minimum bar should be for a reviewer. If 4 people are reviewing, maybe all 4 will assume the others are looking for code quality issues so they will slip right through the review process.
4. The review team should be made aware of which PRs are up to date and which are being kept around for reference purposes only so it is clear what should and should not be merged into the master branch.

**Comment by ams-tschoening** [comment]
No need to make things unnecessary complicated: If reviewers don't react within some reasonable amount of time and things are as clearly requested any necessary like in #17, simply merge.

@igorkwwrd You created the PR, @DavidThi808 approved already, so feel free to merge. Unless reasonable objections are made, I'll do so in the next few days to not keep things stale for any longer.

@NightOwl888 I'll happily make you a member of this org as well if you want to be able to merge at some point.

@klokedm Would be great to have a PR once you see ready and after #17 got merged.

**Comment by klokedm** [comment]
My plan is to test and finalise the changes, clean-up the build scripts to correctly work with Debug & Release scenarios and when I confirm this works (I am cautiously optimistic based on current tests), I plan to do the following:
 - Create another branch from either master or #17, based on the status of the #17 merge;
 - Manually move over the changes required for .NET Core compatibility into the new branch
 - Create a new cross-fork PR from my fork to the ikvm-revived

This will make sure the history is not polluted by my experiments in understanding the compiler logic.

Expecting to have this done by sometime tomorrow or Tuesday, if nothing unexpected pops up.

Am of course open to alternative ways of doing things.

**Comment by ams-tschoening** [comment]
> Create another branch from either master or #17, based on the status of the #17 merge;

Drop a note when you are at that point and the PR has not been merged yet, I will do so then. Sounds like it would make your work easier at least a bit.

**Comment by igorkwwrd** [comment]
#17 merged into the master.

**Comment by klokedm** [comment]
Dear all,

@igorkwwrd - thank you for the merge; I think I am at a point where I will create a new branch from master in my fork and star moving over the changes.

The status is currently as follows:
 - I have temporarily disabled the AWT project. It causes a lot of problems as it cannot target .NET Standard, which means a completely different SDK is targetted. This will need to be supported in the compielr as the reference assemblies are different
 - I have configured the build process to correctly produce the refs folder for the .NET Standard 2.1 (using the KeepCompileContext settings)
 - the compiler currently makes the assumption that it is installed in a folder next to the netstandard2.1, which contains the publisg/refs folder with the appropriate reference assemblies for .NET Standard 2.1
 - the compiler seem to be working, I need to make the build process more robust and prepare some documentation on how to use it so it can be properly tested
 - I will make the compiler auto-include all DLLs in the resolution process that are in the refs folder; it is extremely complicated to correctly build the compile command otherwise
 
As a consequence (until the build files are updated) you will currently need to first build the ikvmc, run the publish target and then build the whole project. This makes sure that the correct reference DLLs are copied to the publish
For convenience there is currently a TestBuild.bat file in the root folder that performs the above steps.
 
As mentioned here: https://github.com/ikvm-revived/ikvm/issues/20#issuecomment-633661789, the ikvm for .NET Core will probably need to target .NET Standard 2.1 (or develop a polyfill for the missing functionality).

Also, I believe the AWT will need to be moved into a seperate project / build as it target a different framework (netcoreapp) and SDK because of Windows Forms.

The next step from my side will be to finalize the build of ikvm so that the whole project builds & executes, then I will start working on porting this to a new branch and executing a PR. I expect to have this done by the end of the week.

**Comment by klokedm** [comment]
A short update, some bad news and some good news. 
First the good news: I’ve tested the IKVM build, it manages to compile a fairl complex Lucene based project and run it, but…
The bad news: while the OpenJDK now target .NET Core / Standard, there was a number of assumptions made in the java part of the implementation on how / what the .NET returns that are no longer true, mostly in the IO related libraries. One example: the FileKey class, used in managing the file locks table used the Stream.Name function to get the filename; however, based on how the stream is constructed, this may return [Unknown], which caused interesting locking exceptions as the JDK assumed different files were actually the same file.
Additionally, some of the APIs called by the OpenJDK compile fine, but return the PlatformNotSupported exception when run in a .NET Core 3.1 context.
I intend to fix the OpenJDK with the errors I find in my current workload and publish a new version + a binary build of the compiler + libraries. If you have any workloads you would like to test with ikvmc, I would appreciate it, if you could test them with the build once provided and let me know of the results.

**Comment by ams-tschoening** [comment]
> The bad news:[...]

Whatever you come across and can't handle on your own, simply create issues here.

> I intend to fix the OpenJDK with the errors I find in my current workload and publish a new version + a binary build of the compiler + libraries.[...]

I have a repo containing different binaries in different branches already. If(!) it makes any sense at all, that might be used to store your results as well and could be transferred to `ikvm-revived`.

https://github.com/ams-ts-ikvm/ikvm-bin

**Comment by klokedm** [comment]
> I have a repo containing different binaries in different branches already. If(!) it makes any sense at all, that might be used to store your results as well and could be transferred to `ikvm-revived`.
> 
> https://github.com/ams-ts-ikvm/ikvm-bin

@ams-tschoening - I think it would make sense to create a binary release and a testing log as part of the repository, that way we could also get some results of the tests.

**Comment by klokedm** [comment]
Another short update from my side:
  - the FileSystem IO has been updated and a windows only solution for the "Append" file mode based on the Microsoft Platform Extensions has been added
  - there have been further runtime compatibility updates to the compiler, specifically the way dynamic runtime compiler resolves methods and fields. Since the .NET framework Token based resolution is sometimes unavaialble, we use the GetMethod approach in those cases
  - there is a problem in the compiled code where it throws MethodAccessException when accessing the public static helper method (see: https://github.com/ikvm-revived/ikvm/issues/21). It seems something has changed either in the emitter logic or in the visibility check logic. If someone has any ideas of the possible causes of the error please comment the above issue.

A binary pre-release is now available here:
https://github.com/klokedm/ikvm/releases/tag/alpha-v01

**Comment by ripenko** [comment]
> Keeping the name is fine.

For me it is also fine.

**Comment by NightOwl888** [comment]
We now have a prerelease that supports .NET Core and .NET Framework https://github.com/ikvm-revived/ikvm/issues/15. Thanks everyone for your participation in making it happen!
